docs/SECURITY.md

```diff
@@ -0,0 +1,185 @@
+# This document outlines protocol-level security mechanisms
+
+## Custody `IChannel` implementation
+
+### Challenge
+
+HORIZONTAL: what is the last known on-chain state of a channel being challenged
+VERTICAL: what is the status of a channel being challenged
+CHANOPEN: a magic number showing a state is the first deposit state
+CHANCLOSE: a magic number showing a state is final
+
+Here is an overview of possible scenarios a channel might be in while being challenged:
+
+|         | CHANOPEN not all joined | CHANOPEN all joined | operatable state | CHANCLOSE |
+| ------- | ----------------------- | ------------------- | ---------------- | --------- |
+| INITIAL | ✅                      | ❌                  | ❌               | ❌        |
+| ACTIVE  | ❌                      | ✅                  | ✅               | ❌        |
+| DISPUTE | ✅                      | ✅                  | ✅               | ❌        |
+| FINAL   | ❌                      | ❌                  | ❌               | ❌        |
+
+As you can see, the channel can not be challenged in a FINAL status or while having CHANCLOSE state.
+
+Let's review each scenario in detail.
+X: an on-chain state the channel is in.
+Y: a state the channel is being challenged with.
+
+General rule is that Y can not be CHANCLOSE. In such case a user should call `close` function.
+Another general rule is that after all checks there are the following on-chain changes:
+
+- `meta.status = DISPUTE`
+- `meta.lastValidState = Y`
+- `meta.challengeExpire = block.timestamp + meta.chan.challenge`
+
+#### INITIAL status
+
+> The main goal: to verify Y is valid and >= X.
+
+```md
+- X is CHANOPEN not all joined:
+  - if (Y is CHANOPEN):
+    - verify Y is a valid CHANOPEN state (has no proof, has CHANOPEN magic number)
+      verify Y has not less signatures than X
+      verify all Y signatures are valid
+      verify all participants that supplied a signature in Y have deposited
+  - else
+    - verify all participants have deposited
+      verify adjudicate(Y, proof)
```

Comment:
Seems like this "branch" contradicts itself by requiring: a) not all joined and b) all participants have deposited. Or deposit and join are different concepts?

Reply:
The `case` statement for this branch states: `X is CHANOPEN not all joined`. The "not all joined" part is a characteristic of State X (can be determined by comparing number of signatures in it to number of participants in a channel). 
On the other hand, the code in question `verify all participants have deposited` references the current "on-chain situation", which may have changed (actually, should have changed as we are in this logical branch), i.e. by some party joining a channel and depositing funds.

I will clarify the fact that "not all joined" references the state (in fact, it references the channel at the moment of that state being added).

Reply:
274ff8ee66e2c2032a8095d37d77f7dc87004c16